src/document-processor.ts

```diff
@@ -0,0 +1,365 @@
+import {attachCommentsToFirstStatement, canDomModuleBeInlined, createDomNodeInsertStatements, filterClone, getCommentsBetween, getNodePathInProgram, insertStatementsIntoProgramBody, serializeNodeToTemplateLiteral} from './document-util';
+import {removeNamespaceInitializers} from './passes/remove-namespace-initializers';
+import {removeToplevelUseStrict} from './passes/remove-toplevel-use-strict';
+import {removeUnnecessaryEventListeners} from './passes/remove-unnecessary-waits';
+import {removeWrappingIIFEs} from './passes/remove-wrapping-iife';
+import {rewriteToplevelThis} from './passes/rewrite-toplevel-this';
+import {ConvertedDocumentUrl, OriginalDocumentUrl} from './urls/types';
+import {UrlHandler} from './urls/url-handler';
+import {isOriginalDocumentUrlFormat} from './urls/util';
+import {replaceHtmlExtensionIfFound} from './urls/util';
+
+/**
+ * Keep a set of elements to ignore when Recreating HTML contents by adding
+ * code to the top of a program.
+ */
+const generatedElementBlacklist = new Set<string|undefined>([
+  'base',
+  'link',
+  'meta',
+  'script',
+]);
+
+/**
+ * An abstract superclass for our document scanner and document converters.
+ */
+export abstract class DocumentProcessor {
+  protected readonly originalUrl: OriginalDocumentUrl;
+  /**
+   * N.B. that this converted url always points to .js, even if this document
+   * will be converted to an HTML file.
+   */
+  protected readonly convertedUrl: ConvertedDocumentUrl;
+  protected readonly urlHandler: UrlHandler;
+  protected readonly conversionSettings: ConversionSettings;
+  protected readonly document: Document;
+  protected readonly program: Program;
+  protected readonly convertedHtmlScripts: ReadonlySet<Import>;
+
+  constructor(
+      document: Document, urlHandler: UrlHandler,
+      conversionSettings: ConversionSettings) {
+    this.conversionSettings = conversionSettings;
+    this.urlHandler = urlHandler;
+    this.document = document;
+    this.originalUrl = urlHandler.getDocumentUrl(document);
+    this.convertedUrl = this.convertDocumentUrl(this.originalUrl);
+    ({program: this.program, convertedHtmlScripts: this.convertedHtmlScripts} =
+         this.prepareJsModule());
+  }
+
+  private isInternalNonModuleImport(scriptImport: Import): boolean {
+    const oldScriptUrl = this.urlHandler.getDocumentUrl(scriptImport.document);
+    const newScriptUrl = this.convertScriptUrl(oldScriptUrl);
+    const isModuleImport =
+        dom5.getAttribute(scriptImport.astNode, 'type') === 'module';
+    const isInternalImport =
+        this.urlHandler.isImportInternal(this.convertedUrl, newScriptUrl);
+    return isInternalImport && !isModuleImport;
+  }
+
+  /**
+   * Creates a single program from all the JavaScript in the current document.
+   * The standard program result can be used for either scanning or conversion.
+   *
+   * TODO: this does a lot of mutation of the program. Could we only do that
```

Comment:
I talked about this with Russel & Justin a bit today, but +1. It would be great if scanning didn't modify the formatted/processed `program` at all, and instead returned data that would then inform the scan results & different conversion steps. 

That way scanning passes could be decoupled/not order-dependent, and conversion passes could be even simpler and more easily configured.